### PR TITLE
Simplify clock_gettime() conditionalization and fallback on macOS

### PR DESCRIPTION
**Pull request by seanm**
Removed the logic in configure.ac (for Darwin only) that detected if the clock_gettime() function exists.  Instead, just use the preprocessor to introspect the macOS SDK and deployment target and call clock_gettime() if they are new enough.

Also replaced the fallback code for older macOS with mach_absolute_time() and gettimeofday(), which are simplier because they do not need setup and teardown carefully balanced and their usage is in one place instead of spread in multiple places in the file.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -201,27 +201,7 @@ AC_CHECK_HEADERS([sys/time.h])
...
-			 osx_10_12_or_later=])
-		if test "x$osx_10_12_or_later" = xyes; then
-			AC_CHECK_FUNCS([clock_gettime], [have_clock_gettime=yes], [have_clock_gettime=])
-		else
-			AC_MSG_NOTICE([clock_gettime() is not available on target OS X version])
-		fi
-	else
-		AC_CHECK_FUNCS([clock_gettime], [have_clock_gettime=yes], [AC_MSG_ERROR([clock_gettime() is required on this platform])])
-	fi
+	AC_CHECK_FUNCS([clock_gettime], [have_clock_gettime=yes], [AC_MSG_ERROR([clock_gettime() is required on this platform])])
```
Comment (tormodvolden) [inline configure.ac:204]: Wouldn't this show an error when building on older macOS?

File: libusb/os/darwin_usb.c [hunk 2]
```diff
@@ -2332,27 +2306,36 @@ static int darwin_handle_transfer_completion (struct usbi_transfer *itransfer) {
   return usbi_handle_transfer_completion (itransfer, darwin_transfer_status (itransfer, tpriv->result));
 }
 
-#if !defined(HAVE_CLOCK_GETTIME)
 void usbi_get_monotonic_time(struct timespec *tp) {
-  mach_timespec_t sys_time;
+// If the SDK is new enough to have the clock_gettime() declaration, and the deplopment target is at least 10.12...
+#if 0 && ((MAC_OS_X_VERSION_MAX_ALLOWED >= 101200) && (MAC_OS_X_VERSION_MIN_REQUIRED >= 101200))
```
Comment (tormodvolden) [inline libusb/os/darwin_usb.c:2311]: #if 0 ?

File: configure.ac (hunk 1 above)
Comment (seanm) [reply configure.ac:204]: Oh maybe.  I'm not familiar with this `.ac` language... from the name, I guess `AC_MSG_ERROR` would pop an error.  I guess I could restore the Darwin check and just not perform this check at all.

File: libusb/os/darwin_usb.c [hunk 3]
```diff
@@ -2332,27 +2306,36 @@ static int darwin_handle_transfer_completion (struct usbi_transfer *itransfer) {
...
-  mach_timespec_t sys_time;
-
-  /* CLOCK_REALTIME represents time since the epoch */
-  clock_get_time (clock_realtime, &sys_time);
-
-  tp->tv_sec  = sys_time.tv_sec;
-  tp->tv_nsec = sys_time.tv_nsec;
-}
+// If the SDK is new enough to have the clock_gettime() declaration, and the deplopment target is at least 10.12...
+#if 0 && ((MAC_OS_X_VERSION_MAX_ALLOWED >= 101200) && (MAC_OS_X_VERSION_MIN_REQUIRED >= 101200))
```
Comment (tormodvolden) [inline libusb/os/darwin_usb.c:2330]: #if 0 here also.

**Comment by tormodvolden** [comment]
It looks like you are doing two potentially independent things here: 1) moving the logic out of configure.ac and into darwin_usb.c and 2) replace the old fallback with mach_absolute_time(). If not two commits, it could at least be stated in the commit message.

File: configure.ac (hunk 1 above)
Comment (tormodvolden) [reply configure.ac:204]: It will print "clock_gettime() is required on this platform" which is false :)

File: libusb/os/darwin_usb.c [hunk 4]
```diff
@@ -2332,27 +2306,36 @@ static int darwin_handle_transfer_completion (struct usbi_transfer *itransfer) {
   return usbi_handle_transfer_completion (itransfer, darwin_transfer_status (itransfer, tpriv->result));
 }
 
-#if !defined(HAVE_CLOCK_GETTIME)
 void usbi_get_monotonic_time(struct timespec *tp) {
-  mach_timespec_t sys_time;
+// If the SDK is new enough to have the clock_gettime() declaration, and the deplopment target is at least 10.12...
```
Comment (tormodvolden) [inline libusb/os/darwin_usb.c:2310]: > If the SDK is new enough to have the clock_gettime()
- isn't that what HAVE_CLOCK_GETTIME will tell us?

Comment (seanm) [reply libusb/os/darwin_usb.c:2310]: They are similar, yes. But compile-time checks are generally preferable to configure-time checks. The latter are an anti-pattern IMNSHO, they are very cross-compile unfriendly.  Why go to the trouble of building a test program when you can just adjust your build with the preprocessor?  Especially when we need a solution that works both with autotools and Xcode.

**Comment by tormodvolden** [comment]
Also bonus points if you explain why gettimeofday() and mach_absolute_time() are better than clock_get_time().

**Comment by seanm** [comment]
> It looks like you are doing two potentially independent things here: 1) moving the logic out of configure.ac and into darwin_usb.c and 2) replace the old fallback with mach_absolute_time(). If not two commits, it could at least be stated in the commit message.

Correct.  My commit message was also WIP, but I've expanded it now.

File: configure.ac (hunk 1 above)
Comment (seanm) [reply configure.ac]: Indeed. I've taken another stab. Does it look more correct now?

**Comment by seanm** [comment]
> Also bonus points if you explain why gettimeofday() and mach_absolute_time() are better than clock_get_time().

Done.

File: libusb/os/darwin_usb.c (hunk 4 above)
Comment (tormodvolden) [reply libusb/os/darwin_usb.c:2310]: And feature checks are preferable to version checks :) but I agree this is not obvious when "cross-compiling" for a different macOS version. I guess configure is great for setting all the overall flags for a big build, but fine-navigation between versions subtleties are easier with the preprocessor.

Theoretically feature checks would be more robust e.g. in case this gets built in a pimped-up build environment where e.g. clock_gettime() could show up supported through compatibility add-ons (à la macports). But we can leave all that thinking to macports :)

File: configure.ac [hunk 5]
```diff
@@ -201,25 +201,7 @@ AC_CHECK_HEADERS([sys/time.h])
...
-			 osx_10_12_or_later=yes],
-			[AC_MSG_RESULT([no])
-			 osx_10_12_or_later=])
-		if test "x$osx_10_12_or_later" = xyes; then
-			AC_CHECK_FUNCS([clock_gettime], [have_clock_gettime=yes], [have_clock_gettime=])
-		else
-			AC_MSG_NOTICE([clock_gettime() is not available on target OS X version])
-		fi
-	else
+	if test "x$backend" != xdarwin; then
```
Comment (tormodvolden) [inline configure.ac:204]: I appreciate the minimum invasiveness, but these two tests could be &&'d for readability.

Comment (seanm) [reply configure.ac:204]: Which two tests?

**Comment by tormodvolden** [comment]
Great.

Typo "wiith" in commit message.

File: configure.ac (hunk 5 above)
Comment (tormodvolden) [reply configure.ac:204]:
```
if test "x$platform" = xposix; then
	dnl check availability of clock_gettime()
	if test "x$backend" != xdarwin; then
```

Comment (seanm) [reply configure.ac:204]: But doesn't the `dnl` statement need to run for the first test but not the second?

File: libusb/os/darwin_usb.c [hunk 6]
```diff
@@ -2332,27 +2306,36 @@ static int darwin_handle_transfer_completion (struct usbi_transfer *itransfer) {
...
+#if ((MAC_OS_X_VERSION_MAX_ALLOWED >= 101200) && (MAC_OS_X_VERSION_MIN_REQUIRED >= 101200))
```
Comment (tormodvolden) [inline libusb/os/darwin_usb.c:2311]: Isn't MAC_OS_X_VERSION_MAX_ALLOWED always >=  MAC_OS_X_VERSION_MIN_REQUIRED? So that only the latter needs checking?

File: configure.ac (hunk 5 above)
Comment (seanm) [reply configure.ac:204]: Oh damn is that a comment?!

Comment (tormodvolden) [reply configure.ac:204]: I think it might get something printed, but if you are not checking for the availability it shouldn't display either.

Comment (seanm) [reply configure.ac]: OK, I've taken another stab.

File: libusb/os/darwin_usb.c (hunk 4 above)
Comment (seanm) [reply libusb/os/darwin_usb.c:2310]: >And feature checks are preferable to version checks :)

The version check is the feature check.  clock_gettime() is documented to have been added in 10.12, thus checking the version is checking for the feature.

>but I agree this is not obvious when "cross-compiling" for a different macOS version

Or any cross-compilation.  Imagine trying to build libusb for iOS from macOS.  Or building for ARM on Intel.  Or building for FreeBSD from macOS. Any test you do on the machine that's building can give the wrong result for the machine you're building *for*.

But I don't want to start a religious argument. :)

File: configure.ac (hunk 5 above)
Comment (tormodvolden) [reply configure.ac]: stackoverflow to the rescue: dnl stands for “Discard to Next Line”  (so pure comment)

File: configure.ac [hunk 7]
```diff
@@ -199,29 +199,9 @@ esac
...
-		fi
-	else
-		AC_CHECK_FUNCS([clock_gettime], [have_clock_gettime=yes], [AC_MSG_ERROR([clock_gettime() is required on this platform])])
-	fi
+dnl check availability of clock_gettime(), except don't bother on Darwin, becuause the result is not used.
+if test "x$platform" = xposix && test "x$backend" != xdarwin; then
+	AC_CHECK_FUNCS([clock_gettime], [have_clock_gettime=yes], [AC_MSG_ERROR([clock_gettime() is required on this platform])])
 
 	if test "x$have_clock_gettime" = xyes; then
 		dnl the clock_gettime() function needs certain clock IDs defined
```
Comment (tormodvolden) [inline configure.ac:207]: Will these certain ID's be needed on newer macOS also? Because now we skipped that test above and $have_clock_gettime is unset.

File: configure.ac (hunk 5 above)
Comment (seanm) [reply configure.ac]: Yeah, found similar. What a weird comment indicator.

File: libusb/os/darwin_usb.c (hunk 4 above)
Comment (tormodvolden) [reply libusb/os/darwin_usb.c:2310]: I am regularly cross-building libusb, e.g. for Windows on Linux. autoconf does the right thing if you've got the details right.

File: configure.ac (hunk 7 above)
Comment (seanm) [reply configure.ac:207]: It looks to be doing a compile-test for `CLOCK_MONOTONIC` and `CLOCK_REALTIME`. Not necessary to perform such a test, the usage of those two constants is also guarded in the `#if ((MAC_OS_X_VERSION_MAX_ALLOWED >= 101200) && (MAC_OS_X_VERSION_MIN_REQUIRED >= 101200))` test.

Comment (tormodvolden) [reply configure.ac:207]: Sounds good. Good you added "becuause the result is not used." but typo in bacuause.

Comment (seanm) [reply configure.ac:207]: Man I suck at typing!

File: libusb/os/darwin_usb.c (hunk 6 above)
Comment (seanm) [reply libusb/os/darwin_usb.c:2311]: Hmm, it generally would be, but I'm not 100% sure it *must* be, so I'd argue to keep it like this.

Comment (tormodvolden) [reply libusb/os/darwin_usb.c:2311]: OK. Just another nitpick: `/* comments */` are standard except inside Windows files.

Comment (tormodvolden) [reply libusb/os/darwin_usb.c:2311]: "deplopment" :) And the ellipses are unneeded. Maybe rephrase to something shorter and less redundant, e.g. "If the SDK declares clock_gettime(), and the deployment target supports it"

From what I understand MAC_OS_X_VERSION_MAX_ALLOWED is the highest version that the SDK knows about, and the target cannot be set higher. But I am fine with keeping both checks because I don't know for sure. LOL, I googled that macro now and this came on top :) https://lists.apple.com/archives/xcode-users/2005/Aug/msg00399.html

**Comment by seanm** [comment]
I think/hope this should be ready to merge now... @tormodvolden ?

**Comment by tormodvolden** [comment]
@seanm Yes, thanks, I think this is good. But I'll leave it for 1.0.27 unless it fixes some observed bug or regression.

**Comment by seanm** [comment]
>@seanm Yes, thanks, I think this is good. But I'll leave it for 1.0.27 unless it fixes some observed bug or regression.

@tormodvolden can we merge this now?

**Comment by mcuee** [comment]
Please help to fix the github warning.

Merging is blocked
The base branch requires all commits to be signed.

**Comment by tormodvolden** [comment]
This PR is on the top of the list in #1094 so it is not forgotten :) I have just waited with pushing anything after the last release while watching to see if something serious comes up and justifies a fixup release. A few important changes were pushed to the repo. As it stands now, I don't think we'll make a fixup release -> discussion in #1094.

mcuee, I don't think we have to worry about the GitHub warning. It only applies if using GitHubs own workflows. We generally pull and rebase patches manually.

**Comment by mcuee** [comment]
> mcuee, I don't think we have to worry about the GitHub warning. It only applies if using GitHubs own workflows. We generally pull and rebase patches manually.

I see. Thanks. Somehow this pops up recently. Can we disable it?

**Comment by tormodvolden** [comment]
I don't know if it can be disabled, but I don't think it is any problem either since we don't want to use GitHub merge anyway.

**Comment by mcuee** [comment]
> I don't know if it can be disabled, but I don't think it is any problem either since we don't want to use GitHub merge anyway.

Not a real problem, just an eye sore to me...
But it is okay if this can not be disabled. I will get used to this.

**Comment by seanm** [comment]
@tormodvolden friendly ping... I think this is ready to merge..

**Comment by neheb** [comment]
Doesn't this fallback to the mach code unconditionally?

**Comment by seanm** [comment]
>Doesn't this fallback to the mach code unconditionally?

I don't believe so.  Can you elaborate on what you think is wrong?

**Comment by neheb** [comment]
I would think it should be:

```
--- a/libusb/libusbi.h
+++ b/libusb/libusbi.h
@@ -534,7 +534,7 @@ static inline void usbi_localize_device_descriptor(struct libusb_device_descript
        desc->bcdDevice = libusb_le16_to_cpu(desc->bcdDevice);
 }
 
-#if defined(HAVE_CLOCK_GETTIME) && !defined(__APPLE__)
+#if defined(HAVE_CLOCK_GETTIME) || ((MAC_OS_X_VERSION_MAX_ALLOWED >= 101200) && (MAC_OS_X_VERSION_MIN_REQUIRED >= 101200))
 static inline void usbi_get_monotonic_time(struct timespec *tp)
 {
        ASSERT_EQ(clock_gettime(CLOCK_MONOTONIC, tp), 0);
diff --git a/libusb/os/darwin_usb.c b/libusb/os/darwin_usb.c
index 3f3a35af..f49b99ca 100644
--- a/libusb/os/darwin_usb.c
+++ b/libusb/os/darwin_usb.c
@@ -304,7 +304,7 @@ static usb_device_t **darwin_device_from_service (struct libusb_context *ctx, io
   const int max_retries = 5;
 
   /* The IOCreatePlugInInterfaceForService function might consistently return
-     an "out of resources" error with certain USB devices the first time we run 
+     an "out of resources" error with certain USB devices the first time we run
      it. The reason is still unclear, but retrying fixes the problem */
   for (int count = 0; count < max_retries; count++) {
     kresult = IOCreatePlugInInterfaceForService(service, kIOUSBDeviceUserClientTypeID,
@@ -2382,11 +2382,8 @@ static int darwin_handle_transfer_completion (struct usbi_transfer *itransfer) {
   return usbi_handle_transfer_completion (itransfer, darwin_transfer_status (itransfer, tpriv->result));
 }
 
+#if ((MAC_OS_X_VERSION_MAX_ALLOWED < 101200) || (MAC_OS_X_VERSION_MIN_REQUIRED < 101200))
 void usbi_get_monotonic_time(struct timespec *tp) {
-/* Check if the SDK is new enough to declare clock_gettime(), and the deployment target is at least 10.12. */
-#if ((MAC_OS_X_VERSION_MAX_ALLOWED >= 101200) && (MAC_OS_X_VERSION_MIN_REQUIRED >= 101200))
-  clock_gettime(CLOCK_MONOTONIC, tp);
-#else
   mach_timebase_info_data_t machTimeBaseInfo;
   mach_timebase_info(&machTimeBaseInfo);
 
@@ -2398,20 +2395,15 @@ void usbi_get_monotonic_time(struct timespec *tp) {
 
   tp->tv_sec = uptimeSeconds;
   tp->tv_nsec = uptimeNanoRemainder;
-#endif
 }
 
 void usbi_get_real_time(struct timespec *tp) {
-/* Check if the SDK is new enough to declare clock_gettime(), and the deployment target is at least 10.12. */
-#if ((MAC_OS_X_VERSION_MAX_ALLOWED >= 101200) && (MAC_OS_X_VERSION_MIN_REQUIRED >= 101200))
-  clock_gettime(CLOCK_REALTIME, tp);
-#else
   struct timeval tv;
   gettimeofday(&tv, NULL);
   tp->tv_sec = tv.tv_sec;
   tp->tv_nsec = tv.tv_usec * NSEC_PER_USEC;
-#endif
 }
+#endif
 
 #if InterfaceVersion >= 550
 static int darwin_alloc_streams (struct libusb_device_handle *dev_handle, uint32_t num_streams, unsigned char *endpoints,
```

I think that's where my confusion came from.

**Comment by seanm** [comment]
> --- a/libusb/libusbi.h
> +++ b/libusb/libusbi.h
> @@ -534,7 +534,7 @@ static inline void usbi_localize_device_descriptor(struct libusb_device_descript
>         desc->bcdDevice = libusb_le16_to_cpu(desc->bcdDevice);
>  }
>  
> -#if defined(HAVE_CLOCK_GETTIME) && !defined(__APPLE__)
> +#if defined(HAVE_CLOCK_GETTIME) || ((MAC_OS_X_VERSION_MAX_ALLOWED >= 101200) && (MAC_OS_X_VERSION_MIN_REQUIRED >= 101200))
>  static inline void usbi_get_monotonic_time(struct timespec *tp)

No, because on macOS `usbi_get_monotonic_time` is not declared here, it's in darwin_usb.c.
